pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package kvcoord
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/rpc"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// rangefeedMuxer is responsible for coordination and management of mux
+// rangefeeds. rangefeedMuxer caches MuxRangeFeed stream per node, and executes
+// each range feed request on an appropriate node.
+type rangefeedMuxer struct {
+	// eventCh receives events from all active muxStreams.
+	eventCh chan *roachpb.MuxRangeFeedEvent
+
+	// Context group controlling execution of MuxRangeFeed calls.
+	g ctxgroup.Group
+
+	// State pertaining to actively executing MuxRangeFeeds.
+	mu struct {
+		syncutil.Mutex
+
+		// terminalErr set when a terminal error occurs.
+		// Subsequent calls to startMuxRangeFeed will return this error.
+		terminalErr error
+
+		// Each call to start new range feed gets a unique ID which is echoed back
+		// by MuxRangeFeed rpc.  This is done as a safety mechanism to make sure
+		// that we always send the event to the correct consumer -- even if the
+		// range feed is terminated and re-established rapidly.
+		nextStreamID int64
+
+		// muxClient contains a nodeID->MuxRangeFeedClient.
+		muxClients map[roachpb.NodeID]*muxClientState
+		// producers maps streamID to the event producer -- data sent back to the
+		// consumer of range feed events.
+		producers map[int64]*channelRangeFeedEventProducer
+	}
+}
+
+type muxClientState struct {
+	client  roachpb.Internal_MuxRangeFeedClient
+	streams util.FastIntSet
+	cancel  context.CancelFunc
+}
+
+func newRangefeedMuxer(g ctxgroup.Group) *rangefeedMuxer {
+	m := &rangefeedMuxer{
+		eventCh: make(chan *roachpb.MuxRangeFeedEvent),
+		g:       g,
+	}
+	m.mu.muxClients = make(map[roachpb.NodeID]*muxClientState)
+	m.mu.producers = make(map[int64]*channelRangeFeedEventProducer)
+
+	m.g.GoCtx(m.demux)
```

Comment:
nit: perhaps this method could be named for its nature of being long-running, for example `demuxLoop` or something like that. Your call. As is, it could also be in charge of demuxing a single event, or something like that.

Reply:
Done.